testing/file_test/file_test_base.cpp

```diff
@@ -71,15 +68,86 @@ auto FileTestBase::TestBody() -> void {
   llvm::errs() << "\nTo test this file alone, run:\n  bazel test "
                << *subset_target << " --test_arg=" << test_file << "\n\n";
 
+  SetTmpAsWorkingDir();
+
   // Load expected output.
-  std::vector<testing::Matcher<std::string>> expected_stdout;
-  std::vector<testing::Matcher<std::string>> expected_stderr;
+  llvm::SmallVector<std::string> test_files;
+  auto cleanup_test_files = llvm::make_scope_exit([&]() {
+    // Remove all the created files. Even if no files are split, there will be a
+    // symlink to remove.
+    for (const auto& file : test_files) {
+      std::filesystem::remove(file);
+    }
+  });
+  llvm::SmallVector<testing::Matcher<std::string>> expected_stdout;
+  llvm::SmallVector<testing::Matcher<std::string>> expected_stderr;
+  ProcessTestFile(test_files, expected_stdout, expected_stderr);
+  if (HasFailure()) {
+    return;
+  }
+
+  // Capture trace streaming, but only when in debug mode.
+  std::string stdout;
+  std::string stderr;
+  llvm::raw_string_ostream stdout_ostream(stdout);
+  llvm::raw_string_ostream stderr_ostream(stderr);
+  bool run_succeeded = RunWithFiles(test_files, stdout_ostream, stderr_ostream);
+  if (HasFailure()) {
+    return;
+  }
+  EXPECT_THAT(!llvm::StringRef(path().filename()).starts_with("fail_"),
+              Eq(run_succeeded))
+      << "Tests should be prefixed with `fail_` if and only if running them "
+         "is expected to fail.";
+
+  // Check results.
+  EXPECT_THAT(SplitOutput(stdout), ElementsAreArray(expected_stdout));
+  EXPECT_THAT(SplitOutput(stderr), ElementsAreArray(expected_stderr));
+}
+
+auto FileTestBase::SetTmpAsWorkingDir() -> void {
+  char* tmpdir = getenv("TEST_TMPDIR");
+  CARBON_CHECK(tmpdir);
+
+  // Run from the tmpdir.
+  std::error_code ec;
+  std::filesystem::current_path(tmpdir, ec);
+  CARBON_CHECK(!ec) << ec.message();
+}
+
+auto FileTestBase::ProcessTestFile(
+    llvm::SmallVector<std::string>& test_files,
+    llvm::SmallVector<testing::Matcher<std::string>>& expected_stdout,
+    llvm::SmallVector<testing::Matcher<std::string>>& expected_stderr) -> void {
   std::ifstream file_content(path());
+  std::ofstream split_file;
+  bool found_content_pre_split = false;
   int line_index = 0;
   std::string line_str;
   while (std::getline(file_content, line_str)) {
-    ++line_index;
     llvm::StringRef line = line_str;
+    if (line.consume_front("// ---")) {
```

Comment:
WDYT about: `// FILE: a.carbon` as the syntax?

Mostly that seems like it might match the other special comment syntaxes and be fairly unsurprising.

Reply:
I believe the current syntax is chosen to match the behavior of the LLVM `split-file` tool.

Reply:
Ah, hadn't seen that tool, and didn't realize it used this syntax. I'm happy with whichever approach others prefer (consistency with `split-file`, or consistency with other special markers).

If no one else cares, TBH, I'd do our own thing unless we ever expect to share the test suite materially. But it's basically a coin toss for me.